Mathlib/CategoryTheory/Retract.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+import Mathlib.CategoryTheory.EpiMono
+
+/-!
+# Retracts
+
+Defines retracts of objects and morphisms.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+/-- An object `X` is a retract of `Y` if there are morphisms `i : X ⟶ Y` and `r : Y ⟶ X` such
+that `i ≫ r = 𝟙 X`. -/
+structure Retract (X Y : C) where
+  /-- `i : X ⟶ Y` -/
+  i : X ⟶ Y
+  /-- `r : Y ⟶ X` -/
+  r : Y ⟶ X
+  /-- `i ≫ r = 𝟙 X` -/
+  retract : i ≫ r = 𝟙 X
+
+namespace Retract
+
+attribute [reassoc (attr := simp)] retract
+
+variable {X Y : C} (h : Retract X Y)
```

Comment:
You could add a definition `Retract.map {X Y : C} (h : Retract X Y) (F : {X Y : C} (h : Retract X Y) (F : C ⥤ D) : Retract (F.obj X) (F.obj Y) where ...`.